pkg/controller/projects/deploykeys/controller.go

```diff
@@ -0,0 +1,274 @@
+	}
+
+	currentState := cr.Spec.ForProvider.DeepCopy()
+	lateInitializeProjectDeployKey(&cr.Spec.ForProvider, dk)
+	isLateInitialized := !compare.Equal(currentState, &cr.Spec.ForProvider)
+
+	cr.Status.AtProvider = v1alpha1.DeployKeyObservation{
+		ID:        &dk.ID,
+		CreatedAt: &metav1.Time{Time: *dk.CreatedAt},
+	}
+
+	cr.Status.SetConditions(xpv1.Available())
+	isUpToDate := isUpToDate(cr, dk)
+
+	return managed.ExternalObservation{
+		ResourceExists:          true,
+		ResourceUpToDate:        isUpToDate,
+		ResourceLateInitialized: isLateInitialized,
+	}, nil
+}
+
+func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.DeployKey)
+
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotDeployKey)
+	}
+
+	if cr.Spec.ForProvider.ProjectID == nil {
+		return managed.ExternalCreation{}, errors.New(errProjectIDMissing)
+	}
+
+	keySecretRef := cr.Spec.ForProvider.KeySecretRef
+
+	namespacedName := kubeTypes.NamespacedName{
+		Namespace: keySecretRef.Namespace,
+		Name:      keySecretRef.Name,
+	}
+
+	secret := &kubeCoreV1.Secret{}
+	err := e.kube.Get(ctx, namespacedName, secret)
+
+	if err != nil {
+		return managed.ExternalCreation{},
+			errors.Wrap(err, errKeyMissing)
+	}
+
+	keyResponse, _, err := e.client.AddDeployKey(
+		*cr.Spec.ForProvider.ProjectID,
+		generateCreateOptions(string(secret.Data[keySecretRef.Key]), &cr.Spec.ForProvider),
+		gitlab.WithContext(ctx),
+	)
+
+	if err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errCreateFail)
+	}
+
+	id := strconv.Itoa(keyResponse.ID)
+	meta.SetExternalName(cr, id)
+
+	return managed.ExternalCreation{ExternalNameAssigned: true}, nil
+}
+
+func (e *external) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+	cr, ok := mg.(*v1alpha1.DeployKey)
+
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errNotDeployKey)
+	}
+
+	if cr.Spec.ForProvider.ProjectID == nil {
+		return managed.ExternalUpdate{}, errors.New(errProjectIDMissing)
+	}
+
+	idString := meta.GetExternalName(cr)
+	id, err := strconv.Atoi(idString)
+
+	if err != nil {
+		return managed.ExternalUpdate{}, errors.Wrap(err, errIDNotAnInt)
+	}
+
+	_, _, er := e.client.UpdateDeployKey(
+		cr.Spec.ForProvider.ProjectID,
+		id,
+		generateUpdateOptions(cr),
+	)
+
+	return managed.ExternalUpdate{}, errors.Wrap(er, errUpdateFail)
+}
+
+func (e *external) Delete(ctx context.Context, mg resource.Managed) error {
+	cr, ok := mg.(*v1alpha1.DeployKey)
+
+	if !ok {
+		return errors.New(errDeleteFail)
+	}
+
+	if cr.Spec.ForProvider.ProjectID == nil {
+		return errors.New(errProjectIDMissing)
+	}
+
+	keyIDString := meta.GetExternalName(cr)
+	keyID, err := strconv.Atoi(keyIDString)
+
+	if err != nil {
+		return errors.Wrap(err, errIDNotAnInt)
+	}
+
+	_, er := e.client.DeleteDeployKey(
+		*cr.Spec.ForProvider.ProjectID,
+		keyID,
+	)
+
+	return errors.Wrap(er, errDeleteFail)
```

Comment:
You can reuse `err` here since it is nil anyway. Use `=` instead of `:=`.

```suggestion
	_, err := e.client.DeleteDeployKey(
		*cr.Spec.ForProvider.ProjectID,
		keyID,
	)

	return errors.Wrap(err, errDeleteFail)
```